rskj-core/src/main/java/co/rsk/net/handler/quota/TxVirtualGasCalculator.java

```diff
@@ -0,0 +1,132 @@
+        this.avgGasPrice = avgGasPrice;
+    }
+
+    /**
+     * Calculates the virtualGas consumed by a transaction taking into account several factors of the transaction and the moment it is being executed
+     *
+     * @param newTx      The tx being executed
+     * @param replacedTx The tx replaced by <code>newTx</code> if any
+     * @return The virtualGas consumed by the provided <code>newTx</code>
+     */
+    double calculate(Transaction newTx, @Nullable Transaction replacedTx) {
+        long txGasLimit = newTx.getGasLimitAsInteger().longValue();
+
+        long newTxNonce = newTx.getNonceAsInteger().longValue();
+        long futureNonceFactor = newTxNonce == accountNonce ? 1 : 2;
+
+        double lowGasPriceFactor = calculateLowGasPriceFactor(newTx);
+
+        double nonceFactor = 1 + NONCE_WEIGHT / (accountNonce + 1);
+
+        double sizeFactor = 1 + newTx.getSize() / SIZE_FACTOR_DIVISOR;
+
+        double replacementFactor = calculateReplacementFactor(newTx, replacedTx);
+
+        double gasLimitFactor = calculateGasLimitFactor(txGasLimit);
+
+        double compositeFactor = futureNonceFactor * lowGasPriceFactor * nonceFactor * sizeFactor * replacementFactor * gasLimitFactor;
+
+        double result = capeResult(txGasLimit, compositeFactor);
+
+        logger.trace("virtualGasConsumed calculation for tx [{}]: result = [{}] (txGasLimit {}, compositeFactor {}, futureNonceFactor {}, lowGasPriceFactor {}, nonceFactor {}, sizeFactor {}, replacementFactor {}, gasLimitFactor {})", newTx.getHash(), result, txGasLimit, compositeFactor, futureNonceFactor, lowGasPriceFactor, nonceFactor, sizeFactor, replacementFactor, gasLimitFactor);
+
+        return result;
+    }
+
+    private double calculateLowGasPriceFactor(Transaction newTx) {
+        long txGasPrice = newTx.getGasPrice().asBigInteger().longValue();
+
+        if (txGasPrice < this.avgGasPrice) {
+            double factor = (this.avgGasPrice - txGasPrice) / (this.avgGasPrice - (double) this.blockMinGasPrice);
+            return 1 + LOW_GAS_PRICE_WEIGH * factor;
+        } else {
+            return 1;
+        }
+    }
+
+    private double calculateReplacementFactor(Transaction newTx, @Nullable Transaction replacedTx) {
+        double newTxGasPrice = newTx.getGasPrice().asBigInteger().doubleValue();
+        double replacementRatio = Optional.ofNullable(replacedTx)
+                .map(Transaction::getGasPrice)
+                .map(rtxGasPrice -> rtxGasPrice.asBigInteger().doubleValue())
+                .map(rtxGasPrice -> newTxGasPrice / rtxGasPrice)
+                .orElse(0.0);
+        return replacementRatio > 0 ? (1 + 1 / replacementRatio) : 1;
+    }
+
+    private double calculateGasLimitFactor(long txGasLimit) {
+        return 1 + GAS_LIMIT_WEIGHT * txGasLimit / this.blockGasLimit;
+    }
+
+    @SuppressWarnings("squid:S1244")
+    // extra security measure
+    private double capeResult(double gasLimit, double factor) {
```

Comment:
This overflow is probably impossible, but it is an extra measure from Security team.
If I finally change to not use primitives as discussed [here](https://github.com/rsksmart/rskj/pull/1749#discussion_r861787148), I would apply your suggestion, otherwise I would leave it like it is.